ethcontract-mock/src/details/mod.rs

```diff
@@ -0,0 +1,911 @@
+
+        let state = self.state.lock().unwrap();
+        Self::ok(&U256::from(state.chain_id))
+    }
+
+    fn transaction_count(&self, mut args: Parser) -> Result<Value, Error> {
+        let address: Address = args.arg();
+        let block: Option<BlockNumber> = args.block_number_opt();
+        args.done();
+
+        let block = block.unwrap_or(BlockNumber::Latest);
+        let state = self.state.lock().unwrap();
+        let transaction_count = match block {
+            BlockNumber::Number(n) if n.as_u64() != state.block => {
+                panic!("mock node does not support returning transaction count for specific block number");
+            }
+            BlockNumber::Earliest => 0,
+            _ => state.nonce.get(&address).copied().unwrap_or(0),
+        };
+        Self::ok(&U256::from(transaction_count))
+    }
+
+    fn gas_price(&self, args: Parser) -> Result<Value, Error> {
+        args.done();
+
+        let state = self.state.lock().unwrap();
+        Self::ok(&U256::from(state.gas_price))
+    }
+
+    fn estimate_gas(&self, mut args: Parser) -> Result<Value, Error> {
+        let _request: CallRequest = args.arg();
+        let _block: Option<BlockNumber> = args.block_number_opt();
+        args.done();
+
+        // TODO:
+        //
+        // We could look up contract's method, match an expectation,
+        // and see if the expectation defines gas price.
+        //
+        // So, for example, this code:
+        //
+        // ```
+        // contract
+        //     .expect_method(signature)
+        //     .with(matcher)
+        //     .gas(100);
+        // ```
+        //
+        // Indicates that call to the method with the given signature
+        // requires 100 gas.
+        //
+        // When estimating gas, we'll check all expectation as if we're
+        // executing a method, but we won't mark any expectation as fulfilled.
+
+        Self::ok(&U256::from(1))
+    }
+
+    fn send_transaction(&self, mut args: Parser) -> Result<Value, Error> {
+        let _request: TransactionRequest = args.arg();
+        args.done();
+
+        // TODO:
+        //
+        // We could support signing if user adds accounts with their private
+        // keys during mock setup.
+
+        panic!("mock node can't sign transactions, use offline signing with private key");
+    }
+
+    fn call(&self, mut args: Parser) -> Result<Value, Error> {
+        let request: CallRequest = args.arg();
+        let block: Option<BlockNumber> = args.block_number_opt();
+
+        let mut state = self.state.lock().unwrap();
+
+        let block = block.unwrap_or(BlockNumber::Latest);
+        match block {
+            BlockNumber::Earliest => {
+                panic!("mock node does not support executing methods on earliest block");
+            }
+            BlockNumber::Number(n) if n.as_u64() != state.block => {
+                panic!("mock node does not support executing methods on non-last block");
```

Comment:
This is something we likely need for GPv2 services. However, seeing as the PR is already quite large, we can leave it for later.

Reply:
In hopes of avoiding this issue for as long as possible we allow specific block number if it's equal to the current top block number. But you're right, we'll have to support this eventually.